flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroSchemaConvertor.java

```diff
@@ -200,4 +272,304 @@ private static DataType handleLogicalTypeSchema(Schema fieldSchema) {
+            case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+            case TIME_WITHOUT_TIME_ZONE:
+                Schema schema = getAvroLogicalType(logicalType);
+                avroSchema = nullable ? nullableSchema(schema) : schema;
+                break;
+            case DECIMAL:
+                Schema decimalSchema = getDecimalSchema(logicalType);
+                avroSchema = nullable ? nullableSchema(decimalSchema) : decimalSchema;
+                break;
+            case ROW:
+                Schema rowSchema = getRowSchema(logicalType, rowName);
+                avroSchema = nullable ? nullableSchema(rowSchema) : rowSchema;
+                break;
+            case MULTISET:
+            case MAP:
+                Schema mapType =
+                        SchemaBuilder.builder()
+                                .map()
+                                .values(
+                                        convertToSchema(
+                                                extractValueTypeToAvroMap(logicalType), rowName));
+                avroSchema = nullable ? nullableSchema(mapType) : mapType;
+                break;
+            case ARRAY:
+                ArrayType arrayType = (ArrayType) logicalType;
+                Schema array =
+                        SchemaBuilder.builder()
+                                .array()
+                                .items(convertToSchema(arrayType.getElementType(), rowName));
+                avroSchema = nullable ? nullableSchema(array) : array;
+                break;
+            default:
+                throw new UnsupportedOperationException(
+                        "Unsupported to derive Schema for type: " + logicalType);
+        }
+        addToLogicalToAvroTypeConversionMap(logicalType, avroSchema);
+        return avroSchema;
+    }
+
+    // -------------------- Helper Methods to handle various schema types --------------------------
+
+    /**
+     * Method to convert {@link RowType} Schema to {@link Schema}.
+     *
+     * @param logicalType {@link LogicalType} Schema to be converted.
+     * @return {@link Schema} of the converted avro Type.
+     */
+    private Schema getRowSchema(LogicalType logicalType, String rowName) {
+        RowType rowType = (RowType) logicalType;
+        List<String> fieldNames = rowType.getFieldNames();
+        // we have to make sure the record name is different in a Schema
+        SchemaBuilder.FieldAssembler<Schema> builder =
+                SchemaBuilder.builder().record(rowName).fields();
+        for (int i = 0; i < rowType.getFieldCount(); i++) {
+            String fieldName = fieldNames.get(i);
+            LogicalType fieldType = rowType.getTypeAt(i);
+            SchemaBuilder.GenericDefault<Schema> fieldBuilder =
+                    builder.name(fieldName)
+                            .type(this.convertToSchema(fieldType, rowName + "_" + fieldName));
+            builder = fieldBuilder.noDefault();
```

Comment:
can this be a part of line 407?

Reply:
We have to assign `fieldBuilder` to `builder`. I don't think merging to 407 will help.